src/klein/_request.py

```diff
@@ -0,0 +1,345 @@
+
+    @attrs(frozen=False)
+    class _State(object):
+        """
+        Internal mutable state for L{HTTPRequestFromIRequest}.
+        """
+
+        _cachedBody = attrib(
+            validator=optional(instance_of(bytes)), default=None, init=False
+        )  # type: bytes
+
+    method  = attrib(validator=instance_of(Text))  # type: Text
+    uri     = attrib(validator=instance_of(URL))  # type: URL
+    headers = attrib(
+        validator=instance_of(FrozenHTTPHeaders)
+    )  # type: FrozenHTTPHeaders
+
+    _body = attrib(validator=validateBody)  # type: Union[bytes, IFount]
+
+    _state = attrib(default=Factory(_State), init=False)  # type: _State
+
+
+    def bodyAsFount(self):
+        # type: () -> IFount
+        if type(self._body) is bytes:
+            return bytesToFount(self._body)
+
+        # assuming: IFount.providedBy(self._body)
+
+        return self._body
+
+
+    def bodyAsBytes(self):
+        # type: () -> bytes
+        if type(self._body) is bytes:
+            return succeed(self._body)
+
+        # assuming: IFount.providedBy(self._body)
+
+        def cache(bodyBytes):
+            # type: (bytes) -> bytes
+            self._state._cachedBody = bodyBytes
+            return bodyBytes
+
+        d = fountToBytes(self._body)
+        d.addCallback(cache)
+        return d
+
+
+
+# Support for L{IRequest}
+
+@implementer(IHTTPRequest)
+@attrs(frozen=True)
+class HTTPRequestFromIRequest(object):
+    """
+    HTTP request.
+
+    This is an L{IHTTPRequest} implementation that wraps an L{IRequest} object.
+
+    This is used by Klein to expose objects from L{twisted.web} to clients
+    while presenting the Klein interface.
+    """
+
+    @attrs(frozen=False)
+    class _State(object):
+        """
+        Internal mutable state for L{HTTPRequestFromIRequest}.
+        """
+
+        _cachedBody = attrib(
+            validator=optional(instance_of(bytes)), default=None, init=False
+        )  # type: bytes
+
+    _request = attrib(validator=provides(IRequest))       # type: IRequest
+
+    _state = attrib(default=Factory(_State), init=False)  # type: _State
+
+
+    @property
+    def method(self):
+        # type: () -> Text
+        return self._request.method.decode("ascii")
+
+
+    @property
+    def uri(self):
+        # type: () -> URL
+        request = self._request
+
+        # This code borrows from t.w.server.Request._prePathURL.
+
+        if request.isSecure():
+            scheme = u"https"
+        else:
+            scheme = u"http"
+
+        netloc = nativeString(request.getRequestHostname())
+
+        port = request.getHost().port
+        if request.isSecure():
+            default = 443
+        else:
+            default = 80
+        if port != default:
+            netloc += u":{}".format(port)
+
+        path = nativeString(request.uri)
+        if path and path[0] == u"/":
+            path = path[1:]
+
+        return URL.fromText(u"{}://{}/{}".format(scheme, netloc, path))
```

Comment:
Rather than mess about with re-joining a bunch of stuff, is it better to create the URL [directly from its components](http://hyperlink.readthedocs.io/en/latest/api.html#hyperlink.URL)?